File: misc/PythonBle/patterns/intense.py
import time

# Loosely based on the "intense" pattern of a 312 box, as documented by Onwrikbaar on
# Joanne's Estim discord (#neostim-chat, 04/06/2024 18:45).
# Channels 1+2 are on constant at 40hz, 130us pulse width. On the 312, the frequency
# could be changed by the MA knob - this is self.freq_hz below. Channels 3+4 use the
# same frequency & pulse width as 1+2, but toggles between the two - switching every 250ms

class intense:
    def __init__(self, ble):
        self.freq_hz = 40 # as would be changed by MA knob (approx 15 - 125)
        self.toggle_period_ms = 250

        self.ble = ble
        self.c3on = False
        self.next_toggle_time_ms = 0

    async def setup(self):
        await self.ble.all_channel_power_level(1000)
        await self.ble.all_channel_frequency(self.freq_hz)
        await self.ble.all_channel_pulse_width(130, 130)

        await self.ble.channel_power_enable(1, self.c3on)
        await self.ble.channel_power_enable(2, self.c3on)

    async def toggle(self):
        self.c3on = not self.c3on
        await self.ble.channel_power_enable(3, self.c3on)
        await self.ble.channel_power_enable(4, not self.c3on)

    async def loop(self):
        time_ms = time.time()*1000.0
        if time_ms > self.next_toggle_time_ms:
            await self.toggle()
            self.next_toggle_time_ms = time_ms + self.toggle_period_ms

File: source/zc95/LuaScripts/intense.lua

-- Loosely based on the "intense" pattern of a 312 box, as documented by Onwrikbaar on
-- Joanne's Estim discord (#neostim-chat, 2024/06/04 18:45).
-- Channels 1+2 are on constant with a configurable frequency and the default pulse width (currently 150us) 
-- Channels 3+4 are the same freq/pulse width, but toggles between the two with a configurable toggle time
-- Using just channels 2+3 and not changing the defaults would be similar-ish to running intense on a 312
-- box with the MA dial in the middle.

_toggle_delay_ms = 250
_freq_1_2 = 40
_freq_3_4 = 40

_last_toggle_time_ms = 0
_chan3_on = false

Config = {
    name = "Intense",
    menu_items = {
        {
            type = "MIN_MAX",
            title = "Toggle delay",
            id = 1,
            min = 100,
            max = 1000,
            increment_step = 50,
            uom = "ms",
            default = _toggle_delay_ms
         },
         {
            type = "MIN_MAX",
            title = "Chan 1&2 freq",
            id = 2,
            min = 10,
            max = 150,
            increment_step = 5,
            uom = "hz",
            default = _freq_1_2
         },
         {
            type = "MIN_MAX",
            title = "Chan 3&4 freq",
            id = 3,
            min = 10,
            max = 150,
            increment_step = 5,
            uom = "hz",
            default = _freq_3_4
         }
    }
}

function Setup(time_ms)

    for chan = 1, 2, 1
    do
        zc.SetFrequency(chan, _freq_1_2)
        zc.ChannelOn(chan)
        zc.SetPower(chan, 1000)
    end

    for chan = 3, 4, 1
    do
        zc.SetFrequency(chan, _freq_3_4)
        zc.SetPower(chan, 1000)
    end
end

function MinMaxChange(menu_id, min_max_val)
    if (menu_id == 1)
    then
        _toggle_delay_ms = min_max_val
    end

    if (menu_id == 2)
    then
        _freq_1_2 = min_max_val
        zc.SetFrequency(1, _freq_1_2)
        zc.SetFrequency(2, _freq_1_2)
    end

    if (menu_id == 3)
    then
        _freq_3_4 = min_max_val
        zc.SetFrequency(3, _freq_3_4)
        zc.SetFrequency(4, _freq_3_4)
    end
end

function Loop(time_ms)
    if (time_ms > _last_toggle_time_ms + _toggle_delay_ms)
    then
        if (_chan3_on == true)
        then
            zc.ChannelOff(3)
            zc.ChannelOn(4)
            _chan3_on = false
        else
            zc.ChannelOn(3)
            zc.ChannelOff(4)
            _chan3_on = true
        end

        _last_toggle_time_ms = time_ms
    end
end
